demos/common/monitors/cpu_monitor.cpp

```diff
@@ -0,0 +1,274 @@
+// Copyright (C) 2019 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "cpu_monitor.h"
+#ifdef WIN32
+#include <tchar.h>
+
+namespace {
+std::size_t getNCores() {
+    SYSTEM_INFO sysinfo;
+    GetSystemInfo(&sysinfo);
+    return sysinfo.dwNumberOfProcessors;
+}
+}
+
+CpuMonitor::CpuMonitor() : nCores{getNCores()}, lastEnabled{false}, historyEnabled{false}, samplesNumber{0} {};
+
+void CpuMonitor::openQuery() {
+    PDH_STATUS status;
+
+    status = PdhOpenQuery(NULL, NULL, &query);
+    if (ERROR_SUCCESS != status) {
+        throw std::runtime_error("PdhOpenQuery() failed");
+    }
+
+    for (std::size_t i = 0; i < nCores; ++i)
+    {
+        TCHAR szCounterName[MAX_PATH];
+        _stprintf_s(szCounterName, sizeof(szCounterName), TEXT("\\Processor(%u)\\%% Processor Time"), i);
+
+        PDH_HCOUNTER counter;
+
+        status = PdhAddCounter(query, szCounterName, 0, &counter);
+        if (ERROR_SUCCESS != status)
+        {
+            throw std::runtime_error("PdhAddCounter() failed");
+        }
+
+        status = PdhSetCounterScaleFactor(counter, -2); // scale counter to [0, 1]
+        if (ERROR_SUCCESS != status)
+        {
+            throw std::runtime_error("PdhSetCounterScaleFactor() failed");
+        }
+
+        coreTimeCounters.push_back(counter);
+    }
+    status = PdhCollectQueryData(query);
```

Comment:
Why do this if you don't read the counters afterwards?

Reply:
Docs say there are some counters that require an extra query before reporting data. But I didn't check if `\Processor(%u)\% Processor Time` is this counter.

Reply:
Could you show me where it says that?

Reply:
https://docs.microsoft.com/en-us/windows/win32/perfctrs/collecting-performance-data

This page also shows that there are zeroes for `\Processor(%u)\% Processor Time` for the first call, so this counter requires two calls.